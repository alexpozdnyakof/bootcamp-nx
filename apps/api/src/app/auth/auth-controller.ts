import { ApiCredentials, ApiSignUp } from '@bootcamp-nx/api-interfaces'
import { Router } from 'express'
import { ResponseWithMessage } from '../response-types'
import { TypedRequest } from '../typed-request'
import { TypedResponse } from '../typed-response'
import { validateEmail } from '../utils'
import AuthService from './auth-service'
import { ApiCredentialsDTO, ApiSignUpDTO } from './credentials'

const AuthController = Router()
const AuthRouterPrefix = 'auth'

const authService = AuthService()
AuthController.post(
	'/sign-in',
	async (
		req: TypedRequest<{ body: ApiCredentials }>,
		res: TypedResponse<ResponseWithMessage>
	) => {
		try {
			const { username, password } = ApiCredentialsDTO.check(req.body)
			const jwtToken = await authService.SignIn({ username, password })

			return res
				.status(200)
				.cookie('refreshToken', jwtToken, {
					httpOnly: true,
					sameSite: 'strict',
				})
				.json({
					code: 200,
					message: 'Authorized',
				})
		} catch (error) {
			console.log(error)
			res.status(401).send({ code: 401, message: error?.message })
		}
	}
)

AuthController.post(
	'/sign-up',
	async (
		req: TypedRequest<{ body: ApiSignUp }>,
		res: TypedResponse<ResponseWithMessage>
	) => {
		try {
			const signUpDTO = ApiSignUpDTO.check(req.body)
			const { password, username } = signUpDTO

			if (!validateEmail(username)) throw new Error('Email is invalid')

			await authService.SignUp(signUpDTO)
			const jwtToken = await authService.SignIn({ username, password })

			res.status(201)
				.cookie('refreshToken', jwtToken, {
					httpOnly: true,
					sameSite: 'strict',
				})
				.json({ code: 201, message: 'Authorized' })
		} catch (error) {
			console.log(error)
			res.status(400).send({ code: 400, message: 'Bad Request' })
		}
	}
)

AuthController.get('/logout', (req, res) => {
	res.clearCookie('refreshToken').end()
})

AuthController.get('/user', (req, res) => {
	if (req.user) res.status(200).send(req.user)
	else res.status(401).send({ error: 'Not Authorized' })
})

export { AuthController, AuthRouterPrefix }
