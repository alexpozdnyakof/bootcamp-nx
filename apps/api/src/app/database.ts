import { Database } from 'sqlite3'

const dbFileName =
	process.env.NODE_ENV === 'test' ? `database-test` : `database`

const database = new Database(`${__dirname}/${dbFileName}.sqlite`, err => {
	if (err) throw new Error(err.message)
})

function migrate() {
	console.log({ __dirname })

	database.exec(
		'CREATE TABLE IF NOT EXISTS projects (id INTEGER PRIMARY KEY AUTOINCREMENT,title VARCHAR(200) NOT NULL,description TEXT NULL)'
	)

	if (process.env.NODE_ENV === 'test') {
		console.log('test')
		addMigrationData()
	}
}

function addMigrationData() {
	console.log({ __dirname })
	database.exec(
		`INSERT OR REPLACE INTO projects VALUES (1, 'ホームページのリニューアル', '新しいフレッシュホームページの制作過程はこちら'), (2, 'チェックアウトフォームのリファクタリング', '新しいチェックアウト フォームでユーザー エクスペリエンスを向上'), (3, '実験プロジェクト', null)`
	)
}

export { database, migrate, addMigrationData }

// 	'CREATE TABLE IF NOT EXISTS projects (id INTEGER PRIMARY KEY AUTOINCREMENT,title VARCHAR(200) NOT NULL,description TEXT NULL)'
// `INSERT OR REPLACE INTO projects VALUES (1, 'ホームページのリニューアル', '新しいフレッシュホームページの制作過程はこちら'), (2, 'チェックアウトフォームのリファクタリング', '新しいチェックアウト フォームでユーザー エクスペリエンスを向上'), (3, '実験プロジェクト', null)`
