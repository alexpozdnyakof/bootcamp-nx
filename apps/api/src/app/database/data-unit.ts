type UniqueId = number

type UpdatedCreatedTime = {
	created: number
	updated: number
}

interface EntityPropsMap {
	task: {
		text: string
		done: boolean
	}
	task_list: {
		title: string
	}
	project: {
		title: string
	}
}

type EntityType = 'task' | 'task_list' | 'project'

type DataUnit<T extends EntityType> = {
	id: UniqueId
	parentId: UniqueId
	type: T
	properties: EntityPropsMap[T] & UpdatedCreatedTime
	children: Array<UniqueId>
}

type EntityUnit = DataUnit<'task'> | DataUnit<'task_list'> | DataUnit<'project'>

type AbstractStore = Array<EntityUnit>

type DataUnitDTO = Pick<
	DataUnit<keyof EntityPropsMap>,
	'parentId' | 'children' | 'type'
> & {
	properties: Omit<
		DataUnit<keyof EntityPropsMap>['properties'],
		'created' | 'updated'
	>
}

type DataUnitUpdateValue = Partial<
	Pick<DataUnitDTO, 'properties' | 'children' | 'parentId'>
> &
	Pick<EntityUnit, 'id'>

export {
	AbstractStore,
	EntityType,
	UniqueId,
	EntityUnit,
	DataUnitDTO,
	DataUnit,
	DataUnitUpdateValue,
	UpdatedCreatedTime,
}
