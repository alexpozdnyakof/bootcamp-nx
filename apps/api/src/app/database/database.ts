import { Database } from 'sqlite3'

import * as fs from 'fs'
import * as path from 'path'

const dbFileName = process.env.NODE_ENV === 'test' ? `test` : `database`

const database = new Database(`${__dirname}/${dbFileName}.sqlite`, err => {
	if (err) throw new Error(err.message)
})
const sqlUrl =
	process.env.NODE_ENV === 'test'
		? path.join(__dirname, '../../sql')
		: path.join(__dirname, '/sql')

function migrate() {
	console.log({ __dirname })

	database.exec(fs.readFileSync(sqlUrl.concat('/projects.sql')).toString())

	if (process.env.NODE_ENV === 'test') {
		console.log('test')
		seeds()
	}
}

function seeds() {
	console.log({ __dirname })
	database.exec(
		fs.readFileSync(sqlUrl.concat('/projects-migration.sql')).toString()
	)
}

export { database, migrate }

// 'CREATE TABLE IF NOT EXISTS projects (id INTEGER PRIMARY KEY AUTOINCREMENT,title VARCHAR(200) NOT NULL,description TEXT NULL)'
// `INSERT OR REPLACE INTO projects VALUES
// (1, 'ホームページのリニューアル',
// '新しいフレッシュホームページの制作過程はこちら'),
// (2, 'チェックアウトフォームのリファクタリング', '新しいチェックアウト フォームでユーザー エクスペリエンスを向上'), (3, '実験プロジェクト', null)`
