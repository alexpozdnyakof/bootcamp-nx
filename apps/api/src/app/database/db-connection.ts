import knex from 'knex'
import knexConfig from '../../knexfile'
const dbConnection = () => {
	const envConfig = knexConfig[process.env.NODE_ENV ?? 'development']
	return knex(envConfig)
}

export default dbConnection()
