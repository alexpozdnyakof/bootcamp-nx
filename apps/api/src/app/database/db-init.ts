import { Database } from 'sqlite3'
import { readFileSync, readdirSync } from 'node:fs'
import { join } from 'path'

function readSqlDirectory(subdir: string): Array<string> {
	const path = join(
		__dirname,
		process.env.NODE_ENV === 'test' ? '../../sql' : '/sql',
		subdir
	)
	return readdirSync(path).map(fileName =>
		readFileSync(join(path, '/', fileName)).toString()
	)
}

function migrate(dbInstance: Database) {
	const migrations = readSqlDirectory('/migrations')
	migrations.map(sqlString => dbInstance.exec(sqlString))
}

function seed(dbInstance: Database) {
	const seeds = readSqlDirectory('/seeds')
	seeds.map(sqlString => dbInstance.exec(sqlString))
}

export function InitSqlite() {
	const isTest = process.env.NODE_ENV === 'test'
	const dbpath = isTest ? ':memory:' : join(__dirname, '/database.sqlite')

	const database = new Database(dbpath, err => {
		if (err) throw new Error(err.message)
	})

	migrate(database)
	if (process.env.NODE_ENV === 'test') {
		seed(database)
	}

	return database
}

export default InitSqlite()
