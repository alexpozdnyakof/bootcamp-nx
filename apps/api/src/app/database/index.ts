export { default as DataModel } from './data-model'
export * from './data-unit'
export { default as Database } from './db'
