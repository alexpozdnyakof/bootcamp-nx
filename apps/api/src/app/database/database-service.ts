import database from './database'

type DbResult<T> = {
	status: 'success' | 'failed'
	result: T
}

export function DatabaseService() {
	return Object.freeze({
		async All<T>(
			query: string,
			params?: Array<string | number | boolean>
		): Promise<T> {
			return new Promise((resolve, reject) => {
				database.all(query, params ?? [], (err, result: T) => {
					if (err) reject(err)
					resolve(result)
				})
			})
		},
		async Get<T>(
			query: string,
			params?: Array<string | number | boolean>
		): Promise<T> {
			return new Promise((resolve, reject) => {
				database.get(query, params ?? [], (err, result: T) => {
					if (err) reject(err)
					resolve(result)
				})
			})
		},
		async Run<T>(
			query: string,
			params?: Array<string | number | boolean>
		): Promise<T> {
			return new Promise((resolve, reject) => {
				database.run(query, params ?? [], (err, result: T) => {
					if (err) reject(err)
					resolve(result)
				})
			})
		},
	})
}

export default DatabaseService()
