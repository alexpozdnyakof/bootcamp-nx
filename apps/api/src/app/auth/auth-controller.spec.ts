import { AuthController } from './auth-controller'
import { json } from 'body-parser'
import express from 'express'
import { database } from '../database'
import request from 'supertest'
describe('Auth Controller', () => {
	const env = process.env
	const App = express()

	const USER_DTO = {
		username: 'test4@gmail.com',
		password: 'password1',
		first_name: 'alex',
		last_name: 'pozdnyakof',
		birthdate: new Date().toISOString(),
	}

	const CREDENTIALS = {
		username: 'test@test.com',
		password: 'password4',
	}

	async function makeUserMigrations() {
		await database.migrate.up({
			name: '20230103123928_create_user_table.js',
		})
		await database.migrate.up({
			name: '20230113101227_create_credential_table.js',
		})
		await database.migrate.up({
			name: '20230113102157_create_user_to_credentials_table.js',
		})
		await database.seed.run({ specific: '00-user.js' })
	}

	beforeAll(async () => {
		App.use(json())
		App.use('/', AuthController)

		await makeUserMigrations()
		process.env.SECRET_KEY = '123qwe123'
	})

	afterAll(() => {
		jest.resetModules()
		process.env = env
	})

	describe('SignIn', () => {
		it('should sign in for existing user with match password', async () => {
			const response = await request(App)
				.post('/sign-in')
				.set('Accept', 'application/json')
				.send(CREDENTIALS)

			const cookies = response.headers['set-cookie'][0].split('=')

			expect(response.status).toBe(200)
			expect(cookies[0]).toBe('refreshToken')
		})
		it('should return error for non-existing user', async () => {
			const response = await request(App)
				.post('/sign-in')
				.set('Accept', 'application/json')
				.send({ ...CREDENTIALS, username: '123@mail.ru' })

			expect(response.status).toBe(401)
		})
		xit('should return error for user with wrong password', async () => {
			const response = await request(App)
				.post('/sign-in')
				.set('Accept', 'application/json')
				.send({ ...CREDENTIALS, password: '123123' })

			expect(response.status).toBe(401)
		})
	})

	describe('SignUp', () => {
		xit('should create new user', async () => {
			const response = await request(App)
				.post('/sign-up')
				.set('Accept', 'application/json')
				.send(USER_DTO)

			const cookies = response.headers['set-cookie'][0].split('=')

			expect(response.status).toBe(201)
			expect(cookies[0]).toBe('refreshToken')
		})
		it('should return error if user already exist', async () => {
			const response = await request(App)
				.post('/sign-up')
				.set('Accept', 'application/json')
				.send({
					username: 'test@test.com',
					password: '123qwe123',
				})
			expect(response.status).toBe(400)
		})
		it('should return error if username not sended', async () => {
			const response = await request(App)
				.post('/sign-up')
				.set('Accept', 'application/json')
				.send({
					password: '123qwe123',
				})
			expect(response.status).toBe(400)
		})
		it('should return error if password not sended', async () => {
			const response = await request(App)
				.post('/sign-up')
				.set('Accept', 'application/json')
				.send({
					username: 'test@test.com',
				})
			expect(response.status).toBe(400)
		})
		it('should return error if email invalid', async () => {
			const response = await request(App)
				.post('/sign-up')
				.set('Accept', 'application/json')
				.send({
					username: 'testtest.com',
					password: '123qwe123',
				})
			expect(response.status).toBe(400)
		})
	})
})
