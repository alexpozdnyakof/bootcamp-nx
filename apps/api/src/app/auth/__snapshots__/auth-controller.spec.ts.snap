// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Controller authenticated routes should return current user 1`] = `
Object {
  "birthdate": "2022-11-29 15:31:37",
  "created": "2022-11-29 15:31:37",
  "first_name": "alex",
  "id": 1,
  "last_name": "pozdnyakof",
  "updated": "2022-11-29 15:31:37",
  "username": "test@test.com",
}
`;
